### PR TITLE
Sp 673 correlation

### PR DESCRIPTION
**Pull request by Coupey**

### CODE REVIEW

File: common/middleware/updateCorrelationId.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
+// checks if there is a keycloak request id in the headers and uses that for correlation ID if found
+const cls = require('cls-hooked')
```
Comment (jonhindle) [inline common/middleware/updateCorrelationId.js:2]: const { getNamespace } = require('cls-hooked')

File: common/middleware/updateCorrelationId.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+const { clsNamespace } = require('../config')
+const { updateMDC } = require('../utils/util')
+
+const updateCorrelationId = (req, res, next) => {
+  const {
+    headers: { 'x-request-id': correlationId },
+  } = req
```
Comment (jonhindle) [inline common/middleware/updateCorrelationId.js:9]: Could be extracted in the declaration:
`const updateCorrelationId = ( { headers: { 'x-request-id': correlationId } }, res, next) => {`

File: common/middleware/updateCorrelationId.test.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  let req
+  let res
+  beforeEach(() => {
+    req = {
+      headers: {
+        'x-auth-token': 'THX1138',
+      },
+    }
+  })
+
```
Comment (jonhindle) [inline common/middleware/updateCorrelationId.test.js:31]: This should go in.
```
afterEach(() => {
  jest.resetAllMocks()
})
```

File: common/middleware/updateCorrelationId.test.js [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  test('should not change correlation ID', done => {
+    updateCorrelationId(req, res, done)
+    expect(cls.getNamespace).not.toHaveBeenCalled()
+  })
+
+  test('should update correlation ID', done => {
+    req.headers['x-request-id'] = 'NCC-1701'
+    updateCorrelationId(req, res, done)
+    expect(cls.getNamespace).toHaveBeenCalled()
+    expect(updateMDC).toHaveBeenCalled()
```
Comment (jonhindle) [inline common/middleware/updateCorrelationId.test.js:41]: I think this line is unnecessary.

File: common/utils/util.js [hunk 5]
```diff
@@ -1,4 +1,6 @@
+const cls = require('cls-hooked')
```
Comment (jonhindle) [inline common/utils/util.js:1]: Request just what we need.
`const { getNamespace } = require('cls-hooked')`

File: common/utils/util.js [hunk 6]
```diff
@@ -126,6 +128,16 @@ const getActionText = ({ description = '', intervention }, interventionList) =>
     ? interventionList.find(({ uuid }) => uuid === intervention).shortDescription
     : description
 
+const getCorrelationId = () => {
+  const { correlationId = '' } = cls.getNamespace(clsNamespace)
+  return correlationId
+}
```
Comment (jonhindle) [inline common/utils/util.js:134]: Could be a one liner. Still readable I think?
```
const getCorrelationId = () => getNamespace(clsNamespace).correlationId || ''
```

Both versions though don't cater for `getNamespace` returning undefined. I'm guessing thats not possible, but thought I'd mention :-)

File: common/utils/util.js [hunk 7]
```diff
@@ -126,6 +128,16 @@ const getActionText = ({ description = '', intervention }, interventionList) =>
...
+
+const updateMDC = (mdcDataKey, mdc) => {
+  const thisNamespace = cls.getNamespace(clsNamespace)
+  thisNamespace.set(mdcDataKey, mdc)
+}
```
Comment (jonhindle) [inline common/utils/util.js:139]: Could be a single line if you think its still readable.
```
const updateMDC = (mdcDataKey, mdc) => getNamespace(clsNamespace).set(mdcDataKey, mdc)
```

File: mockServer/app.js [hunk 8]
```diff
@@ -14,7 +14,8 @@ const getFile = (directory, file, res) => {
 
 function createMockAPI() {
   app.use('*', (req, res, next) => {
```
Comment (jonhindle) [inline mockServer/app.js:16]: Maybe a break with whats gone before but:
```
app.use('*', ({ method, originalUrl, headers }, res, next) => {
```

File: common/middleware/updateCorrelationId.js [hunk 9]
```diff
@@ -0,0 +1,17 @@
+// checks if there is a keycloak request id in the headers and uses that for correlation ID if found
+const { getNamespace } = require('cls-hooked')
+const { clsNamespace } = require('../config')
+const { updateMDC } = require('../utils/util')
+
+const updateCorrelationId = ({ headers: { 'x-request-id': correlationId } }, res, next) => {
+  if (correlationId) {
+    const mdcNamespace = getNamespace(clsNamespace)
+    const mdc = mdcNamespace.get('MDC')
```
Comment (jonhindle) [inline common/middleware/updateCorrelationId.js:9]: Worth 'MDC' being a constant? I don't think we gain anything but I'll always ask the question 😬

Comment (Coupey) [reply common/middleware/updateCorrelationId.js:9]: Yeah. No.

**Pull request merged**